partiql-lang/src/main/kotlin/org/partiql/lang/eval/ThunkAsync.kt

```diff
@@ -0,0 +1,613 @@
+/*
+ * Copyright 2019 Amazon.com, Inc. or its affiliates.  All rights reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License").
+ *  You may not use this file except in compliance with the License.
+ * A copy of the License is located at:
+ *
+ *      http://aws.amazon.com/apache2.0/
+ *
+ *  or in the "license" file accompanying this file. This file is distributed on an "AS IS" BASIS,
+ *  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the License for the specific
+ *  language governing permissions and limitations under the License.
+ */
+
+package org.partiql.lang.eval
+
+import com.amazon.ionelement.api.MetaContainer
+import kotlinx.coroutines.flow.flow
+import kotlinx.coroutines.flow.toList
+import org.partiql.errors.ErrorBehaviorInPermissiveMode
+import org.partiql.errors.ErrorCode
+import org.partiql.errors.Property
+import org.partiql.lang.ast.SourceLocationMeta
+import org.partiql.lang.ast.StaticTypeMeta
+import org.partiql.lang.domains.staticType
+import org.partiql.lang.types.StaticTypeUtils.isInstance
+
+/**
+ * A thunk with no parameters other than the current environment.
+ *
+ * See https://en.wikipedia.org/wiki/Thunk
+ *
+ * @param TEnv The type of the environment.  Generic so that the legacy AST compiler and the new compiler may use
+ * different types here.
+ */
+internal typealias ThunkAsync<TEnv> = suspend (TEnv) -> ExprValue
```

Comment:
File is basically the same as `Thunk.kt`. Notable changes include
- Making any functions/lambda/types async (i.e. adding `suspend`)
- Converting sequence constructors that make calls to async functions use coroutine `Flow`s